polars/polars-core/src/series/arithmetic/borrowed.rs

```diff
@@ -406,140 +415,33 @@ fn coerce_time_units<'a>(
 }
 
 #[cfg(feature = "dtype-struct")]
-pub fn _struct_arithmetic<F: FnMut(&Series, &Series) -> Series>(
-    s: &Series,
+pub fn struct_arithmetic_try_op<F: FnMut(&Series, &Series) -> PolarsResult<Series>>(
+    lhs: &Series,
     rhs: &Series,
     mut func: F,
-) -> Series {
-    let s = s.struct_().unwrap();
-    let rhs = rhs.struct_().unwrap();
-    let s_fields = s.fields();
+) -> PolarsResult<Series> {
+    let lhs = lhs.struct_().unwrap_or_else(|_| unreachable!());
```

Comment:
What's the benefit of this over `unwrap`?

Reply:
I think it was a marker for myself while reading this, basically saying this is only called with struct dtypes.

This is one of my biggest gripes with polars codebase 🙂 When staring at a particular piece of code you have no idea what assumptions are currently being used here, you see a bunch of unwraps and have no idea which of them CAN potentially trigger and which cannot ever happen, which are unhandled errors and which are completely unreachable. You have to scan all usages of a particular thing trying to infer it.

I think a lot of these cases you can convert to traits/errors or just restructure the code. E.g.:
- This function should probably take `&StructChunked, &StructChunked` and then no unwraps are needed here
- And upstream instead of matching `(lhs.dtype(), rhs.dtype()) = (Struct(_), Struct(_))` you could instead do `(lhs.struct_(), rhs.struct_()) = (Some(_), Some(_))`, so no unwraps are needed again. No implicit assumptions either.

Reply:
Right. Maybe just put an `unwrap` with a comment why we can unwrap or `expect("impl error")`?

Reply:
> This is one of my biggest gripes with polars codebase 

I agree we can do better here. All unwraps should be considered asserts, but it isn't always the case yet.